File: Tree/Binary_Search_Trees/BinarySearchTree.java
class BinarySearchTree
{
    node head;

    static class node
    {
        int data;
        node left;
        node right;

        node(int d)
        {
            data = d;
            left = null;
            right = null;
        }
    }

    public void Insert(int value)
    {
        node temp = new node(value);

        if (head == null)
            head = temp;
        else
        {
            node current;
            current = head;

            while (current != null)
            {
                if (value < current.data)
                {
                    if (current.left != null)
                        current = current.left;
                    else
                    {
                        current.left = temp;
                        return ;
                    }
                }
                else
                {
                    if (current.right != null)
                        current = current.right;
                    else
                    {
                        current.right = temp;
                        return ;
                    }
                }
            }
        }
    }

    public void Search(int value)
    {
        node current;
        current = head;

        while (current != null)
        {
            if (value < current.data)
                current = current.left;
            else if (value > current.data)
                current = current.right;
            else
            {
                System.out.println("Element " + value + " Found");
                return ;
            }
        }

        System.out.println("Element " + value + " not Found");
    }

    public int Min_Value(node head)
    {
        while (head.left != null)
            head = head.left;
        return head.data;
    }

    public node Delete_Key(node head, int value)
    {
        if (head == null)
            return head;

        if (value < head.data)
            head.left = Delete_Key(head.left, value);
        else if (value > head.data)
            head.right = Delete_Key(head.right, value);
        else
        {
            if (head.left == null)
                return head.right;
            else if (head.right == null)
                return head.left;

            head.data = Min_Value(head.right);
            head.right = Delete_Key(head.right, head.data);
        }

        return head;
    }

    public void Delete(int value)
    {
        head = Delete_Key(head, value);
    }

    public static void main(String[] args)
    {
        BinarySearchTree BST = new BinarySearchTree();

        BST.Insert(5);
        BST.Insert(7);
        BST.Insert(9);
        BST.Insert(8);
        BST.Insert(6);
        BST.Insert(4);

        BST.Search(9);
        BST.Search(2);

        BST.Delete(7);
        BST.Delete(5);
        BST.Delete(4);

        BST.Search(9);
        BST.Search(2);
        BST.Search(5);
        BST.Search(6);
    }
}

/* Output

Element 9 Found
Element 2 not Found
Element 9 Found
Element 2 not Found
Element 5 not Found
Element 6 Found

*/

File: Binary_Search_Trees/go/BinarySearchTree.go
package main

import (
	"fmt"
)

// BinarySearchTree is the struct implementation of BST
type BinarySearchTree struct {
	Root *BinarySearchTreeNode
}

// BinarySearchTreeNode represents the node of binary search tree
type BinarySearchTreeNode struct {
	Data  int
	Left  *BinarySearchTreeNode
	Right *BinarySearchTreeNode
}

// Insert inserts the data in the  BinarySearchTree
func (t *BinarySearchTree) Insert(data int) {

	if t.Root == nil {
		t.Root = &BinarySearchTreeNode{Data: data}
		return
	}
	insert(t.Root, data)
}

// Delete deletes the node with value equal to data from the BST
func (t *BinarySearchTree) Delete(data int) {
	delete(t.Root, data)
}

// Search searches the BST for given data
func (t *BinarySearchTree) Search(data int) bool {
	return search(t.Root, data)
}

// MinValue returns the minimum value in the BST
func (t *BinarySearchTree) MinValue() int {
	return minValue(t.Root)
}

// MaxValue returns the minimum value in the BST
func (t *BinarySearchTree) MaxValue() int {
	return maxValue(t.Root)
}

// Inorder prints the BST in Inorder traversal
func (t *BinarySearchTree) Inorder() {
	inorder(t.Root)
}

func insert(n *BinarySearchTreeNode, data int) {

	if n.Data > data {
		if n.Left == nil {
			n.Left = &BinarySearchTreeNode{
				Data:  data,
				Left:  nil,
				Right: nil,
			}
			return
		}
		insert(n.Left, data)
	} else {
		if n.Right == nil {
			n.Right = &BinarySearchTreeNode{
				Data:  data,
				Left:  nil,
				Right: nil,
			}
			return
		}
		insert(n.Right, data)
	}
}

func delete(n *BinarySearchTreeNode, data int) *BinarySearchTreeNode {

	if n == nil {
		return nil
	}

	if n.Data == data {
		if n.Left == nil && n.Right == nil {
			return nil
		}
		if n.Left == nil {
			return n.Right
		} else if n.Right == nil {
			return n.Left
		}
		min := minValue(n.Right)
		n.Data = min
		n.Right = delete(n.Right, min)
		return n
	}

	if data < n.Data {
		n.Left = delete(n.Left, data)
	} else {
		n.Right = delete(n.Right, data)
	}
	return n
}

func search(n *BinarySearchTreeNode, data int) bool {

	if n == nil {
		return false
	}
	if n.Data == data {
		return true
	} else if n.Data > data {
		return search(n.Left, data)
	} else {
		return search(n.Right, data)
	}
}

func minValue(root *BinarySearchTreeNode) int {
	if root.Left == nil {
		return root.Data
	}
	return minValue(root.Left)
}

func maxValue(root *BinarySearchTreeNode) int {
	if root.Right == nil {
		return root.Data
	}
	return maxValue(root.Right)
}

func inorder(root *BinarySearchTreeNode) {
	if root == nil {
		return
	}
	inorder(root.Left)
	fmt.Printf("%d ", root.Data)
	inorder(root.Right)
}

func main() {
	// Create a Binary Search Tree
	var tree *BinarySearchTree = &BinarySearchTree{}
	// Insert elements in BST
	tree.Insert(5)
	tree.Insert(4)
	tree.Insert(6)
	tree.Insert(3)
	tree.Insert(7)
	tree.Insert(2)
	tree.Insert(8)
	tree.Insert(1)
	tree.Insert(9)
	// Inorder Traversal
	fmt.Print("Inorder traversal of BST is: ")
	tree.Inorder()
	// Search if 4 is present (should print true)
	fmt.Println("\nIs 4 present in BST? ", tree.Search(4))
	// Delete 4 from BST and Search if 4 is present (should print false)
	tree.Delete(4)
	fmt.Println("Is 4 present in BST? ", tree.Search(4))
	// Inorder Traversal after deleting 4
	fmt.Print("Inorder traversal of BST is: ")
	tree.Inorder()
	// Print min value in BST
	fmt.Printf("\nMinimum value in BST is: %d\n", tree.MinValue())
	// Print max value in BST
	fmt.Printf("Maximum value in BST is: %d\n", tree.MaxValue())
}

/*
	Output:
	Inorder traversal of BST is: 1 2 3 4 5 6 7 8 9
	Is 4 present in BST?  true
	Is 4 present in BST?  false
	Inorder traversal of BST is: 1 2 3 5 6 7 8 9
	Minimum value in BST is: 1
	Maximum value in BST is: 9
*/
